lib/topological_inventory/orchestrator/worker.rb

```diff
@@ -1,97 +1,100 @@
+      def initialize(api_base_url: ENV["API_URL"], collector_definitions_file: ENV["CDF"])
+        @api_base_url = api_base_url
+        @collector_definitions_file = collector_definitions_file || TopologicalInventory::Orchestrator.root.join("config/collector_definitions.yaml")
+        @logger = ManageIQ::Loggers::Container.new
       end
 
       def run
         loop do
-          Signal.trap("TERM") { break }
-          collector_definitions.each do |source_type, endpoint_info|
-            endpoint_info.each do |endpoint_role, options|
-              resolve_collector_type(source_type, endpoint_role, options)
-            end
-          end
+          make_openshift_match_database
+
           sleep 10
         end
       end
 
-      def resolve_collector_type(source_type, endpoint_role, options)
-      end
+      private
 
-      def create_collectors_for_new_sources
-        sources.each do |source|
-          create_objects_for_source(source) unless have_deployment_for_source?(source)
-        end
-      end
-      
-      def remove_collectors_for_deleted_sources
-        collector_deployments.each do |deployment|
-          unless sources.include?(source_for_deployment(deployment))
-            @object_manager.delete_deployment_config(deployment.metadata.name)
-            @object_manager.delete_secret("#{deployment.metadata.name}-secrets")
-          end
-        end
+      def digest(object)
+        require 'digest'
+        Digest::SHA1.hexdigest(Marshal.dump(object))
       end
 
-      def have_deployment_for_source?(source)
-        collector_deployments.detect do |deployment|
-          source_for_deployment(deployment) == source["uid"]
-        end
-      end
+      def make_openshift_match_database
+        hash = {}
+        expected = collectors_from_database(hash)
+        current  = collectors_from_openshift
 
-      def create_objects_for_source(source)
-        @object_manager.create_deployment_config(deployment_name_for_source(source)) do |d|
-          container = d[:spec][:template][:spec][:containers].first
-          container[:env] << collector_container_environment(source)
-          container[:image] = "#{ENV["MY_NAMESPACE"]}/topological-inventory-collector-openshift:latest"
-        end
-      end
+        logger.info("Checking...")
 
-      private
+        (current - expected).each { |i| remove_openshift_objects_for_source(i) }
+        (expected - current).each { |i| create_openshift_objects_for_source(i, hash[i]) }
 
-      def collector_container_environment(source)
+        logger.info("Checking... complete.")
       end
 
-      def secret_name_for_source(source)
-        "#{deployment_name_for_source(source)}-secrets"
-      end
 
-      def deployment_name_for_source(source)
-        "topological-inventory-collector-source-#{source["id"]}"
+      ### API STUFF
+      def each_endpoint
+        return enum_for(:each_endpoint) unless block_given?
```

Comment:
Oh I misread the caller. The block is being sent to collect, not `each_endpoint`.

That said, this still seems harder to read to me. I'm still not really sure how this works after reading the `enum_for` documentation...

Reply:
`enum_for` returns an Enumerator object that will pass the yielded values to the caller.

Reply:
Okay, but I still don't see why this method has to account for being called with a block if we are not going to do that. Were dealing with two cases when we only have one.

Simply, why not make this method return an array rather than yielding?

Reply:
I think what @carbonin is saying is, why do the enumerator dance when this method can just return an Array (i.e. name the method `endpoints` instead of `each_endpoint`

Reply:
Oh, I think the array is a bad idea as far as memory is concerned.  I'd rather only keep the information that we want for as long as it is needed.  This way we can process each resource as we fetch them from the API rather than fetching everything up front then looping over the data.